src/engine/threaded_engine_perdevice.cc

```diff
@@ -87,21 +92,34 @@ class ThreadedEnginePerDevice : public ThreadedEngine {
         int nthread = gpu_worker_nthreads_;
         int dev_id = ctx.dev_id;
         if (is_copy) {
+          auto ptr =
```

Comment:
Note that these are destroying ThreadPool objects, joining the threads in those pools, etc, which is more of a performance issue than this